### DISCUSSION TITLE
[markdown] open a link with a new tab

### DISCUSSION THREAD

**Discussion by inkyusa** [squidfunk/mkdocs-material#3660] [Q&A] [answered]
Hi, thanks for the great work.
I am new to mkdocs and markdown and just wondering how to open a link with a new tab.
I created a github page using mkdocs-material and tried to do that with `[link](url){:target="_blank"}` or `[link](url){:target="\_blank"}` but didn't work.

Only possible workaround for me is using HTML such that 
```html
<a href="http://google.com/" target="_blank">Hello, google!</a>
```
but this way somehow messed up context and I prefer to have a nice concise form: using `[link]` within a sentence and referencing with `[link]: http://google.com`.

It seems I missed something or `mkdocs.yaml` setup issue?
Best

**Comment by squidfunk** [comment, marked as answer]
You probably forgot to activate the Attribute Lists extension:
https://squidfunk.github.io/mkdocs-material/setup/extensions/python-markdown/#attribute-lists

**Reply by inkyusa** [reply]
Thanks for your prompt reply.
Yes, I missed `attr_list`
More concretely, I needed to add the following into `mkdocs.yml`
```
markdown_extensions:
  - attr_list
```
and

`[my-link]{:target="_blank"}` within a sentence and ` [my-link]: http://google.com` for referencing. (Note that there are two leading spaces).

Hope this helps someone who may face this kind of greeting.
Cheers,

**Comment by squidfunk** [comment]
The [built-in privacy plugin](https://squidfunk.github.io/mkdocs-material/setup/ensuring-data-privacy/#built-in-privacy-plugin) now allows to automatically [add `target=_blank` to all external links](https://squidfunk.github.io/mkdocs-material/setup/ensuring-data-privacy/#+privacy.external_links_attr_map), so they open in new tabs.

**Reply by TheNickest** [reply]
*Reserved for insiders/sponsors..

**Reply by thomas-mc-work** [reply]
Couldn't find it mentioned in the linked docs. May it be that this feature has moved somewhere else?

**Reply by thomas-mc-work** [reply]
Found it: https://squidfunk.github.io/mkdocs-material/plugins/privacy/?h=external#config.links

**Reply by kamilkrzyskow** [reply]
Yes, @thomas-mc-work 
it was removed from the Setup page, but the option is still available for sponsors and is mentioned in the plugin configuration
- https://squidfunk.github.io/mkdocs-material/plugins/privacy/#config.links_attr_map

**Comment by JakubAndrysek** [comment]
I have created the plugin to do that automatically. Plugin opens all outgoing links in a new tab.
Look at the plugin: [mkdocs-open-in-new-tab](https://github.com/JakubAndrysek/mkdocs-open-in-new-tab).

Just install the plugin using pip from [PyPI](https://pypi.org/project/mkdocs-open-in-new-tab/):

```bash
pip install mkdocs-open-in-new-tab
```

And add one line of code to your `mkdocs.yml`

```yaml
plugins:
  - search
  - open-in-new-tab
```

Demo: https://newtab.kubaandrysek.cz.

**Reply by River-Mochi** [reply]
thank you your plugin was easy to set up and worked right away.
I just added plugin and one line to my mkdocs.yml  👍 

could not get the code suggested above to work right for my situation. your plugin  ended an hour of frustration :)

**Reply by patppuccin** [reply]
Worked like a charm, thanks a ton.

**Reply by oscgonfer** [reply]
:clap: thank you!

**Reply by yebo29** [reply]
Thank you so much for this plugin. Did exactly what I needed very easily.

**Reply by quantfreedom** [reply]
best day ever!!!! thanks!

**Reply by nkudupudi** [reply]
More power to you mate ! Thank you.

**Reply by studentwiki** [reply]
SUPER! thank you!

**Reply by squidfunk** [reply]
Albeit I understand, that you all might be thankful for the plugin, I'm kindly referring to [our policy to use reactions](https://squidfunk.github.io/mkdocs-material/contributing/#before-commenting) for this kind of feedback. Please remember: __everybody in this discussion will get a notification__ when you post a comment.

> Does your comment add value to the conversation? Is it constructive and respectful to our community and us maintainers? Could you just use a [ reaction](https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) instead?

Thus, if your comment doesn't add any value to the conversation, please just use 🚀👍🎉❤️. Thank you for your understanding ☺️

**Reply by RobertoPrevato** [reply]
Beware that using `target="_blank"` for all anchors is not good for accessibility, especially for users using screen readers.

**Comment by TheJazzDev** [comment]
@JakubAndrysek Looks like this plugin you created only works for python correct?
I'm trying to integrate it into my NextJS app.

**Reply by JakubAndrysek** [reply]
Hi, I have designed this plugin for Mkdocs usage, but it is only a piece of javascript injected into the final webpage.
There is the [js code](https://github.com/JakubAndrysek/mkdocs-open-in-new-tab/blob/main/open_in_new_tab/js/open_in_new_tab.js) - taken from [jekyllcodex.org](https://jekyllcodex.org/without-plugin/new-window-fix/). 

Look at the [documentation](https://newtab.kubaandrysek.cz/#testing) for more details.

**Reply by TheJazzDev** [reply]
Thanks mate

**Comment by thesuperzapper** [comment]
Here is a [MkDocs "hook"](https://www.mkdocs.org/user-guide/configuration/#hooks) that injects a ["Python-Markdown Extension"](https://python-markdown.github.io/extensions/api/) which makes all external links use `"target="_blank" and rel="noopener"`:

```python
from typing import List
from urllib.parse import urlparse

import markdown
from mkdocs.config.defaults import MkDocsConfig


class ExternalLinksTreeProcessor(markdown.treeprocessors.Treeprocessor):
    """
    Adds target="_blank" and rel="noopener" to external links.
    """

    def __init__(self, md, ignore_domains: List[str]):
        super().__init__(md)
        # convert to set for faster lookup
        self.ignore_domains = set(ignore_domains)

    def run(self, root):
        for element in root.iter():
            if element.tag == "a":
                href = element.get("href", "")

                # parse the href and check if it's an absolute URL with http or https scheme
                parsed_url = urlparse(href)

                if parsed_url.scheme in ["http", "https"]:
                    # remove port, if present
                    domain = parsed_url.netloc.split(":")[0]

                    # skip if the domain is in the ignore list
                    if domain in self.ignore_domains:
                        continue

                    element.set("target", "_blank")
                    element.set("rel", "noopener")


class ExternalLinksExtension(markdown.Extension):
    def __init__(self, **kwargs):
        self.config = {
            "ignore_domains": [[], "List of domains to ignore"],
        }
        super(ExternalLinksExtension, self).__init__(**kwargs)

    def extendMarkdown(self, md: markdown.Markdown):
        ignore_domains = self.getConfig("ignore_domains", [])
        if not isinstance(ignore_domains, list):
            raise ValueError("'ignore_domains' config must be a list")

        md.treeprocessors.register(
            ExternalLinksTreeProcessor(md, ignore_domains), "external_links", -1000
        )


def on_config(config: MkDocsConfig, **kwargs):
    # a list of domains to ignore
    # WARNING: requires re-running `mkdocs serve` when changed
    IGNORE_DOMAINS = ["example.com"]

    # inject the markdown extension
    config.markdown_extensions.append(
        ExternalLinksExtension(ignore_domains=IGNORE_DOMAINS)
    )
    return config
```

You can exclude specific domains by updating the `IGNORE_DOMAINS` variable.

All you need to do is put this file in your MkDocs repo, and reference it under `hooks` your `mkdocs.yml`, for example:

```yaml
hooks:
  - path/to/external_links_md_extension.py
```

**Reply by squidfunk** [reply]
Thanks for sharing! Note that this will only influence links that are part of Markdown — other links, like for example those in the footer or navigation will be left untouched. The [`privacy` plugin's external links](https://squidfunk.github.io/mkdocs-material/setup/ensuring-data-privacy/#external-links) feature allows to catch all of them.

**Comment by z-aki** [comment]
wrote a plugin to simplify this https://github.com/z-aki/mkdocs-plugin-open-external-links-in-new-tab/
